message-transport/pkg/transport/metrics/consumer.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package metrics
+
+import (
+	"github.com/projectriff/riff/message-transport/pkg/message"
+	"github.com/projectriff/riff/message-transport/pkg/transport"
+	"encoding/json"
+	"log"
+	"time"
+)
+
+const consumerSource = "consumer"
+
+// ConsumerAggregateMetric represents the reception of a number of message from a topic by a function in a time interval.
+type ConsumerAggregateMetric struct {
+	Topic    string
+	Function string
+	Pod      string
+	Interval time.Duration
+	Count    int32
+}
+
+// NewConsumer decorates the given delegate to send consumer metrics for the given function and pod to the given topic
+// using the given metrics producer. The given pod can be any unique identifier for the pod which will use this consumer
+// and can be pod instance-specific (that is, it need not carry across when a pod is restarted).
+func NewConsumer(delegate transport.Consumer, function string, pod string, metricsTopic string, metricsProducer transport.Producer) *consumer {
+	return &consumer{
+		function:        function,
```

Comment:
I wonder if we should really make the term 'function' appear here. After all, these Producer/Consumer abstractions have nothing specific to functions

Reply:
We could use the term 'consumer group', but I would like to avoid the Kafka bias that might be implied by that term.

Do you have another application of these metrics in mind other than to (Riff) functions?

Reply:
I think "consumer group" is getting more and more accepted as a more-than-kafka thing, eg redis streams, etc

Reply:
Let's go with 'consumer group' and rename if this seems too broken when we support Rabbit MQ, for example.

Reply:
Renamed.